### PR TITLE
Add .entitiesAt() to Plots.Rectangle

### PR DESCRIPTION
**Pull request by rcchen**
image

Still need to add `.entitiesWithin(range)`

Partially addresses #1959 (resolves for Plots.Rectangle)

### CODE REVIEW

File: src/plots/rectanglePlot.ts [hunk 1]
```diff
@@ -213,6 +213,45 @@ export module Plots {
...
+     * Gets the Entities at a particular Point.
+     * 
+     * @param {Point} p
+     * @param {PlotEntity[]} 
+     */
+    public entitiesAt(p: Point) {
+      var results: PlotEntity[] = [];
+      var entities = this.entities();
+      var xScale: Scale<any, number> = this.x().scale;
+      var yScale: Scale<any, number> = this.y().scale;
```
Comment (aicioara) [inline src/plots/rectanglePlot.ts]: No need to type it

File: src/plots/rectanglePlot.ts [hunk 2]
```diff
@@ -213,6 +213,45 @@ export module Plots {
...
+      for (var i = 0; i < entities.length; i++) {
```
Comment (aicioara) [inline src/plots/rectanglePlot.ts]: `entitites.forEach()` to be consistent with bar plot

File: src/plots/rectanglePlot.ts [hunk 3]
```diff
@@ -213,6 +213,45 @@ export module Plots {
...
+        var entity = entities[i];
+        var x = xScale.scale(this.x().accessor(entity.datum, entity.index, entity.dataset));
+        var y = yScale.scale(this.y().accessor(entity.datum, entity.index, entity.dataset));
+        var x2 = x;
+        if (this.x2()) {
+          x2 = xScale.scale(this.x2().accessor(entity.datum, entity.index, entity.dataset));
+        } else {
+          var xRangeBand = (<Plottable.Scales.Category> xScale).rangeBand();
```
Comment (aicioara) [inline src/plots/rectanglePlot.ts]: This is not coherent. Why is the xScale a category scale if there is no `x2()` accessor? Is it always the case? What if the user just forgot to specify the `x2()`, but uses a linear scale? Is that not a scenario?

File: src/plots/rectanglePlot.ts [hunk 4]
```diff
@@ -213,6 +213,45 @@ export module Plots {
...
+          x -= xRangeBand / 2;
+          x2 += xRangeBand / 2;
+        }
+        var y2 = y;
+        if (this.y2()) {
+          y2 = yScale.scale(this.y2().accessor(entity.datum, entity.index, entity.dataset));
+        } else {
+          var yRangeBand = (<Plottable.Scales.Category> yScale).rangeBand();
```
Comment (aicioara) [inline src/plots/rectanglePlot.ts]: Again, is this always the case?

File: src/plots/rectanglePlot.ts [hunk 5]
```diff
@@ -213,6 +213,45 @@ export module Plots {
...
+          y -= yRangeBand / 2;
+          y2 += yRangeBand / 2;
+        }
+        if (((x < x2 && p.x > x && p.x < x2) || (x > x2 && p.x > x2 && p.x < x)) &&
+            ((y < y2 && p.y > y && p.y < y2) || (y > y2 && p.y > y2 && p.y < y))) {
```
Comment (aicioara) [inline src/plots/rectanglePlot.ts]:
```typescript
if (Math.min(x, x2) < p.x && p.x < Math.max(x, x2))
```
cleaner.

Comment (aicioara) [inline src/plots/rectanglePlot.ts]: Should it be `<=` instead of `<` ?

File: test/plots/rectanglePlotTests.ts [hunk 6]
```diff
@@ -39,6 +39,44 @@ describe("Plots", () => {
       VERIFY_CELLS((<any> rectanglePlot)._renderArea.selectAll("rect"));
       svg.remove();
     });
+
+    it("retrieves the correct entity under a point", () => {
+      var xScale = new Plottable.Scales.Linear();
+      var yScale = new Plottable.Scales.Linear();
+      var svg = TestMethods.generateSVG(SVG_WIDTH, SVG_HEIGHT);
+      var dataset = new Plottable.Dataset(DATA);
+      var plot = new Plottable.Plots.Rectangle()
```
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Can we parameterize this plot?

File: test/plots/rectanglePlotTests.ts [hunk 7]
```diff
@@ -39,6 +39,44 @@ describe("Plots", () => {
...
+        .x((d) => d.x, xScale).x2((d) => d.x2)
+        .y((d) => d.y, yScale).y2((d) => d.y2)
+        .addDataset(dataset).renderTo(svg);
+      var entities = (<Plottable.Plots.Rectangle<number, number>> plot).entitiesAt({ x: 140, y: 140 });
```
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: should not need casting if parameterized above.

File: test/plots/rectanglePlotTests.ts [hunk 8]
```diff
@@ -39,6 +39,44 @@ describe("Plots", () => {
...
+      assert.lengthOf(entities, 1, "a single entity is retrieved at (140, 140)");
+      assert.strictEqual(entities[0].index, 2, "entity retrieved is at index 2");
+      svg.remove();
+    });
+
+    it("retrieves correct entities under a point", () => {
```
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Duplicate test name. makes it hard to debug. Rename or collapse the two

File: test/plots/rectanglePlotTests.ts [hunk 9]
```diff
@@ -39,6 +39,44 @@ describe("Plots", () => {
...
+      var xScale = new Plottable.Scales.Linear();
+      var yScale = new Plottable.Scales.Linear();
+      var svg = TestMethods.generateSVG(SVG_WIDTH, SVG_HEIGHT);
+      var dataset = new Plottable.Dataset([
+        { x: 1, y: 1, x2: 3, y2: 3 },
+        { x: 2, y: 2, x2: 4, y2: 4 }
```
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Can we have a dataset where we also test x > x2 and y > y2?

File: test/plots/rectanglePlotTests.ts [hunk 10]
```diff
@@ -39,6 +39,44 @@ describe("Plots", () => {
...
+      ]);
+      var plot = new Plottable.Plots.Rectangle()
+        .x((d) => d.x, xScale).x2((d) => d.x2)
+        .y((d) => d.y, yScale).y2((d) => d.y2)
+        .addDataset(dataset).renderTo(svg);
+      var entities = (<Plottable.Plots.Rectangle<number, number>> plot).entitiesAt({ x: 150, y: 150 });
```
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Same, should not require casting

File: test/plots/rectanglePlotTests.ts [hunk 11]
```diff
@@ -39,6 +39,44 @@ describe("Plots", () => {
...
+      assert.lengthOf(entities, 2, "two entities are retrieved at (150, 150)");
+      entities = (<Plottable.Plots.Rectangle<number, number>> plot).entitiesAt({ x: 200, y: 50 });
```
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: same

File: test/plots/rectanglePlotTests.ts [hunk 12]
```diff
@@ -39,6 +39,44 @@ describe("Plots", () => {
...
+      assert.lengthOf(entities, 1, "one entity is retrieved at (200, 50)");
+      assert.strictEqual(entities[0].index, 1, "entity retrieved is at index 1");
+      entities = (<Plottable.Plots.Rectangle<number, number>> plot).entitiesAt({ x: 50, y: 200 });
```
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Same

File: src/plots/rectanglePlot.ts (hunk 1 above)
Comment (rcchen) [inline src/plots/rectanglePlot.ts]: I need it for the casting to `Scales.Category`

File: test/plots/rectanglePlotTests.ts (hunk 8 above)
Comment (rcchen) [inline test/plots/rectanglePlotTests.ts]: The first one is a single entity retrieved; the second one is multiple entities.

**Comment by bluong** [comment]
Taking over this pull request

**Comment by aicioara** [comment]
Whoa... what's up with the merge? why are all the commits above part of this PR?

File: src/plots/rectanglePlot.ts [hunk 13]
```diff
@@ -231,6 +231,30 @@ export module Plots {
       return this;
     }
 
+    /**
+     * Gets the Entities at a particular Point.
+     *
+     * @param {Point} p
+     * @param {PlotEntity[]} 
+     */
+    public entitiesAt(p: Point) {
```
Comment (aicioara) [inline src/plots/rectanglePlot.ts]: type it. Let's avoid single letter variables.

File: src/plots/rectanglePlot.ts [hunk 14]
```diff
@@ -231,6 +231,30 @@ export module Plots {
       return this;
     }
 
+    /**
+     * Gets the Entities at a particular Point.
```
Comment (aicioara) [inline src/plots/rectanglePlot.ts]: We use capitals in comments when we refer to a class, but there is no class called `Entities`

File: src/plots/rectanglePlot.ts [hunk 15]
```diff
@@ -231,6 +231,30 @@ export module Plots {
...
+     *
+     * @param {Point} p
```
Comment (aicioara) [inline src/plots/rectanglePlot.ts]: Since this is a public method, let's document the parameter more thoroughly.

File: src/plots/rectanglePlot.ts [hunk 16]
```diff
@@ -231,6 +231,30 @@ export module Plots {
...
+     * @param {PlotEntity[]} 
+     */
+    public entitiesAt(p: Point) {
+      var results: PlotEntity[] = [];
+      var attrToProjector = this._generateAttrToProjector();
+      this.entities().forEach((entity) => {
```
Comment (aicioara) [inline src/plots/rectanglePlot.ts]: this.entities().filter

File: test/plots/rectanglePlotTests.ts [hunk 17]
```diff
@@ -3,17 +3,17 @@
...
-    ];
+  var SVG_WIDTH = 300;
+  var SVG_HEIGHT = 300;
+  var DATA = [
+    { x: 0, y: 0, x2: 1, y2: 1 },
+    { x: 1, y: 1, x2: 2, y2: 2 },
+    { x: 2, y: 2, x2: 3, y2: 3 },
+    { x: 3, y: 3, x2: 4, y2: 4 },
+    { x: 4, y: 4, x2: 5, y2: 5 }
+  ];
```
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Put these inside the `RectanglePlot` describe. The outer most describe is used for display in the test page only.

**Comment by aicioara** [comment]
See my comments at the top of this PR as well.

File: src/plots/rectanglePlot.ts (hunk 14 above)
Comment (bluong) [inline src/plots/rectanglePlot.ts]: `Entity` is a type however

Comment (aicioara) [inline src/plots/rectanglePlot.ts]: But Entities is not. Plus it's not even returning an Entity, it's returning a PlotEntity.

File: test/plots/rectanglePlotTests.ts (hunk 6 above)
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: No need.

File: test/plots/rectanglePlotTests.ts [hunk 18]
```diff
@@ -40,6 +39,44 @@ describe("Plots", () => {
...
+      var svg = TestMethods.generateSVG(SVG_WIDTH, SVG_HEIGHT);
+      var dataset = new Plottable.Dataset([
+        { x: 1, y: 1, x2: 3, y2: 3 },
+        { x: 4, y: 2, x2: 2, y2: 4 }
+      ]);
+      var plot = new Plottable.Plots.Rectangle()
+        .x((d) => d.x, xScale).x2((d) => d.x2)
+        .y((d) => d.y, yScale).y2((d) => d.y2);
+      plot.addDataset(dataset).renderTo(svg);
+      var entities = plot.entitiesAt({ x: 150, y: 150 });
```
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: These are magic numbers. Let's instead use the scale to scale the values to pixels. TBH, I have no idea which point this represents, but assuming it is (2, 2), the LOC would look like
```typescript
var entities = plot.entitiesAt({x: xScale.scale(2), y: yScale.scale(2)})
```

File: test/plots/rectanglePlotTests.ts [hunk 19]
```diff
@@ -40,6 +39,44 @@ describe("Plots", () => {
...
+      assert.lengthOf(entities, 2, "two entities are retrieved at (150, 150)");
```
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Can we check which are the 2 entities returned. Either by checking their index, or otherwise.

File: test/plots/rectanglePlotTests.ts [hunk 20]
```diff
@@ -40,6 +39,46 @@ describe("Plots", () => {
...
+      plot.addDataset(dataset).renderTo(svg);
+      var entities = plot.entitiesAt({ x: xScale.scale(2), y: xScale.scale(2) });
+      assert.lengthOf(entities, 2, "two entities when querying a point in both");
+      assert.strictEqual(entities[0].index, 0, "entity retrieved is at index 0");
+      assert.strictEqual(entities[1].index, 1, "entity retrieved is at index 1");
+      entities = plot.entitiesAt({ x: xScale.scale(4), y: yScale.scale(4) });
+      assert.lengthOf(entities, 1, "one entity retrieved");
+      assert.strictEqual(entities[0].index, 1, "entity retrieved is at index 1");
+      entities = plot.entitiesAt({ x: xScale.scale(1), y: yScale.scale(1) });
+      assert.lengthOf(entities, 1, "one entity retrieved");
```
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Specify which entity. This message and the message on line 74 are the same. When the dev sees this message, it is not clear which assert failed.

File: test/plots/rectanglePlotTests.ts [hunk 21]
```diff
@@ -40,6 +39,46 @@ describe("Plots", () => {
...
+    it("retrieves the correct entity under a point", () => {
+      var xScale = new Plottable.Scales.Linear();
+      var yScale = new Plottable.Scales.Linear();
+      var svg = TestMethods.generateSVG(SVG_WIDTH, SVG_HEIGHT);
+      var dataset = new Plottable.Dataset(DATA);
+      var plot = new Plottable.Plots.Rectangle()
+        .x((d) => d.x, xScale).x2((d) => d.x2)
+        .y((d) => d.y, yScale).y2((d) => d.y2);
+      plot.addDataset(dataset).renderTo(svg);
+      var entities = plot.entitiesAt({ x: 140, y: 140 });
```
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Apply the same scaling changes here

File: test/plots/rectanglePlotTests.ts [hunk 22]
```diff
@@ -40,6 +39,46 @@ describe("Plots", () => {
...
+      assert.lengthOf(entities, 1, "a single entity is retrieved at (140, 140)");
```
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: amend this assert as well

File: test/plots/rectanglePlotTests.ts [hunk 23]
```diff
@@ -40,6 +39,46 @@ describe("Plots", () => {
...
+      var plot = new Plottable.Plots.Rectangle()
+        .x((d) => d.x, xScale).x2((d) => d.x2)
+        .y((d) => d.y, yScale).y2((d) => d.y2);
+      plot.addDataset(dataset).renderTo(svg);
+      var entities = plot.entitiesAt({ x: xScale.scale(2), y: xScale.scale(2) });
+      assert.lengthOf(entities, 2, "two entities when querying a point in both");
+      assert.strictEqual(entities[0].index, 0, "entity retrieved is at index 0");
+      assert.strictEqual(entities[1].index, 1, "entity retrieved is at index 1");
+      entities = plot.entitiesAt({ x: xScale.scale(4), y: yScale.scale(4) });
+      assert.lengthOf(entities, 1, "one entity retrieved");
```
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: same

File: test/plots/rectanglePlotTests.ts [hunk 24]
```diff
@@ -40,6 +39,46 @@ describe("Plots", () => {
...
+      var yScale = new Plottable.Scales.Linear();
+      var svg = TestMethods.generateSVG(SVG_WIDTH, SVG_HEIGHT);
+      var dataset = new Plottable.Dataset(DATA);
+      var plot = new Plottable.Plots.Rectangle()
+        .x((d) => d.x, xScale).x2((d) => d.x2)
+        .y((d) => d.y, yScale).y2((d) => d.y2);
+      plot.addDataset(dataset).renderTo(svg);
+      var entities = plot.entitiesAt({ x: 140, y: 140 });
+      assert.lengthOf(entities, 1, "a single entity is retrieved at (140, 140)");
+      assert.strictEqual(entities[0].index, 2, "entity retrieved is at index 2");
```
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Semantics: `second (or third) entity was retrieved`

**Comment by crmorford** [comment]
Cat x Cat http://jsfiddle.net/1h335t30/3/
Time x Cat http://jsfiddle.net/1h335t30/4/
Numeric x Cat http://jsfiddle.net/1h335t30/5/
Numeric x Numeric http://jsfiddle.net/1h335t30/7/
Cat x Numeric http://jsfiddle.net/1h335t30/8/
also testing clicking through when rectangles overlap, clicking empty space, etc...just wanted to give an example for each major case.
